lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,363 @@
+			// Note: using any other error constructor (e.g. BadParameter)
+			// might lead to relogin attempt and a completely obscure
+			// error message
+			return trace.BadParameter("%q is a directory, but the recursive option was not passed", c.srcPaths[i])
+		}
+		fileInfos[i] = fi
+	}
+
+	for i, fi := range fileInfos {
+		dstPath := c.dstPath
+		if dstIsDir || fi.IsDir() {
+			dstPath = path.Join(dstPath, fi.Name())
+		}
+
+		if fi.IsDir() {
+			if err := c.transferDir(ctx, dstPath, c.srcPaths[i], fi); err != nil {
+				return trace.Wrap(err)
+			}
+		} else {
+			if err := c.transferFile(ctx, dstPath, c.srcPaths[i], fi); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// transferDir transfers a directory
+func (c *Config) transferDir(ctx context.Context, dstPath, srcPath string, srcFileInfo os.FileInfo) error {
+	err := c.dstFS.Mkdir(dstPath)
+	if err != nil && !errors.Is(err, os.ErrExist) {
+		return trace.Wrap(err)
+	}
+
+	if err := c.dstFS.Chmod(dstPath, srcFileInfo.Mode()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	infos, err := c.srcFS.ReadDir(srcPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, info := range infos {
+		dstSubPath := path.Join(dstPath, info.Name())
+		lSubPath := path.Join(srcPath, info.Name())
+
+		if info.IsDir() {
+			if err := c.transferDir(ctx, dstSubPath, lSubPath, info); err != nil {
+				return trace.Wrap(err)
+			}
+		} else {
+			if err := c.transferFile(ctx, dstSubPath, lSubPath, info); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	// set modification and access times last so creating sub dirs/files
+	// doesn't update the times
+	if c.opts.PreserveAttrs {
+		err := c.dstFS.Chtimes(dstPath, getAtime(srcFileInfo), srcFileInfo.ModTime())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// transferFile transfers a file
+func (c *Config) transferFile(ctx context.Context, dstPath, srcPath string, srcFileInfo os.FileInfo) error {
+	srcFile, err := c.srcFS.Open(srcPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer srcFile.Close()
+
+	dstFile, err := c.dstFS.Create(dstPath, uint64(srcFileInfo.Size()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer dstFile.Close()
+
+	// write to canceler first so if the context is canceled the transferring
+	// can stop immediately
+	var writer io.Writer
+	canceler := &cancelWriter{
+		ctx: ctx,
+	}
+	// if a progress writer was set, write file transfer progress
+	if c.ProgressWriter != nil {
+		progressBar := newProgressBar(srcFileInfo.Size(), srcFileInfo.Name(), c.ProgressWriter)
+		writer = io.MultiWriter(canceler, dstFile, progressBar)
+	} else {
+		writer = io.MultiWriter(canceler, dstFile)
+	}
+
+	n, err := io.Copy(writer, srcFile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if n != srcFileInfo.Size() {
+		return trace.Errorf("short write: written %v, expected %v", n, srcFileInfo.Size())
+	}
+
+	err = c.dstFS.Chmod(dstPath, srcFileInfo.Mode())
```

Comment:
Is there no way to create the file with the correct mode to begin with? What does OpenSSH sftp do?

Reply:
Updated fork so that destination files and dirs can be created with source modes.